File: app/src/main/java/ru/dimagor555/stocks/data/local/stock/StockModelMapper.java
package ru.dimagor555.stocks.data.local.stock;

import ru.dimagor555.stocks.data.model.stock.Stock;
import ru.dimagor555.stocks.data.model.stock.StockCompanyInfo;
import ru.dimagor555.stocks.data.model.stock.StockPrice;

import javax.inject.Inject;

public class StockModelMapper {
    @Inject
    public StockModelMapper() {
    }

    public StockModel toModel(Stock stock) {
        StockCompanyInfo companyInfo = stock.getCompanyInfo();
        StockPrice price = stock.getPrice();
        return new StockModel(stock.getTicker(),
                companyInfo.getCompanyName(),
                companyInfo.getCompanySiteUrl(),
                companyInfo.getLogoUrl(),
                price.getCurrPriceInCents(),
                price.getPreviousClosePriceInCents(),
                price.getPriceTime(),
                stock.isFavourite());
    }

    public Stock fromModel(StockModel model) {
        return new Stock(model.getTicker(),
                model.getCompanyName(),
                model.getCompanySiteUrl(),
                model.getLogoUrl(),
                model.getCurrPriceInCents(),
                model.getPreviousClosePriceInCents(),
                model.getPriceTime(),
                model.isFavourite());
    }
}

File: app/src/main/java/ru/dimagor555/stocks/data/local/stock/StockModelMapper.kt
package ru.dimagor555.stocks.data.local.stock

import ru.dimagor555.stocks.data.model.stock.Stock
import ru.dimagor555.stocks.data.model.stock.StockCompanyInfo
import javax.inject.Inject

class StockModelMapper @Inject constructor(
    private val priceMapper: StockPriceModelMapper
) {
    fun toStockBaseModel(stock: Stock, priceId: Int): StockBaseModel {
        val companyInfo = stock.companyInfo
        return StockBaseModel(
            stock.ticker,
            companyInfo.companyName,
            companyInfo.companySiteUrl,
            companyInfo.logoUrl,
            stock.isFavourite,
            priceId
        )
    }

    fun toStockPriceModel(stock: Stock, priceId: Int = 0): StockPriceModel {
        return priceMapper.toModel(stock.price, priceId)
    }

    fun fromStockModel(model: StockModel): Stock {
        val companyInfo = with(model.stockBaseModel) {
            StockCompanyInfo(companyName, companySiteUrl, logoUrl)
        }
        val price = priceMapper.fromModel(model.stockPriceModel)
        return Stock(
            model.stockBaseModel.ticker,
            companyInfo,
            price,
            model.stockBaseModel.favourite
        )
    }
}